### PR TITLE
Default destination suggestions during install via CLI

### PR DESCRIPTION
**Pull request by prafullbhosale**
CLI prompts for a destination if none is given and there is no defaultDestination; offers suggestion as a default value

Eg: 
```
> libman install jquery
Default provider [unpkg]: <enter>
Destination [wwwroot\lib\jquery]: <enter>
```

```
> libman install D:\abc\libs\myscripts\ -p filesystem
Destination [wwwroot\lib\myscripts]: <enter>
```

### CODE REVIEW

File: src/LibraryManager.Contracts/ILibraryCatalog.cs [hunk 1]
```diff
@@ -43,5 +43,11 @@ public interface ILibraryCatalog
         /// <param name="cancellationToken">A token that allows the search to be cancelled.</param>
         /// <returns>The library identifier of the latest released version.</returns>
         Task<string> GetLatestVersion(string libraryId, bool includePreReleases, CancellationToken cancellationToken);
+
+        /// <summary>
+        /// Gets the suggested destination path for the library.
+        /// </summary>
+        /// <param name="library"></param>
+        string GetSuggestedDestination(ILibrary library);
```
Comment (raquelsa) [inline src/LibraryManager.Contracts/ILibraryCatalog.cs]: shouldn't this be a provider method?

File: src/LibraryManager/Providers/FileSystem/FileSystemCatalog.cs [hunk 2]
```diff
@@ -136,6 +136,25 @@ public Task<CompletionSet> GetLibraryCompletionSetAsync(string value, int caretP
             return library;
         }
 
+        public string GetSuggestedDestination(ILibrary library)
+        {
+            if (library != null && library is FileSystemLibrary fileSystemLibrary)
```
Comment (raquelsa) [inline src/LibraryManager/Providers/FileSystem/FileSystemCatalog.cs]: Do we need anything for the file system? I guess you can just return empty string.

File: src/LibraryManager.Contracts/ILibraryCatalog.cs (hunk 1 above)
Comment (prafullbhosale) [reply src/LibraryManager.Contracts/ILibraryCatalog.cs]: Moved it to `IProvider`

File: src/LibraryManager/Providers/FileSystem/FileSystemProvider.cs [hunk 3]
```diff
@@ -177,6 +177,30 @@ public ILibraryCatalog GetCatalog()
...
 
+        /// <summary>
+        /// Returns the last valid filename part of the path which identifies the library.
+        /// </summary>
+        /// <param name="library"></param>
+        /// <returns></returns>
+        public string GetSuggestedDestination(ILibrary library)
+        {
+            if (library != null && library is FileSystemLibrary fileSystemLibrary)
+            {
```
Comment (ToddGrun) [inline src/LibraryManager/Providers/FileSystem/FileSystemProvider.cs]: nit: extra newline

File: test/libman.Test/TestInputReader.cs [hunk 4]
```diff
@@ -15,5 +15,15 @@ public string GetUserInput(string fieldName)
         {
             return Inputs[fieldName];
         }
+
+        public string GetUserInputWithDefault(string fieldName, string defaultValue)
+        {
+            if (Inputs.ContainsKey(fieldName))
```
Comment (ToddGrun) [inline test/libman.Test/TestInputReader.cs]: >ContainsKey [](start = 23, length = 11)

I'd use TryGetValue instead that way you don't do the lookup twice

File: test/LibraryManager.Test/Providers/Cdnjs/CdnjsProviderTest.cs [hunk 5]
```diff
@@ -172,6 +172,22 @@ public void Cleanup()
             Assert.AreEqual("LIB018", result.Errors[0].Code);
         }
 
+        [TestMethod]
+        public void GetSuggestedDestination()
+        {
```
Comment (ToddGrun) [inline test/LibraryManager.Test/Providers/Cdnjs/CdnjsProviderTest.cs]: nit: another extra newline

File: LibraryManager.sln [hunk 6]
```diff
@@ -57,8 +57,9 @@ Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "ShimGenerator", "ShimGenera
 	EndProjectSection
 EndProject
 Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "AppHostRestorerProject", "AppHostRestorerProject", "{E9F67355-871B-4104-B887-0BC6D0A04B65}"
-EndProject
-Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "AppHostRestorerProject", "tools\ShimGenerator\AppHostRestorerProject\AppHostRestorerProject.csproj", "{CE6EDDD4-AFBD-4FCF-89FC-30A0CD034642}"
+	ProjectSection(SolutionItems) = preProject
+		tools\ShimGenerator\AppHostRestorerProject\AppHostRestorerProject.csproj = tools\ShimGenerator\AppHostRestorerProject\AppHostRestorerProject.csproj
+	EndProjectSection
```
Comment (jimmylewis) [inline LibraryManager.sln]: This looks like a solution file, not a project in the solution?

File: src/libman/Commands/InstallCommand.cs [hunk 7]
```diff
@@ -3,6 +3,7 @@
 
 using System;
 using System.Collections.Generic;
+using System.IO;
```
Comment (ToddGrun) [inline src/libman/Commands/InstallCommand.cs]: OK, I'm blind, which bit of code needs this?

File: LibraryManager.sln (hunk 6 above)
Comment (prafullbhosale) [reply LibraryManager.sln]: That is correct and is what I intended. 
This file is used by the msbuild magic to generate the shims and fails to restore when the solution restore runs.

File: src/libman/Commands/InstallCommand.cs (hunk 7 above)
Comment (prafullbhosale) [reply src/libman/Commands/InstallCommand.cs]: Line 115 `Path.Combine()`

Comment (ToddGrun) [reply src/libman/Commands/InstallCommand.cs]: Yep, there it is. Thanks!

---
In reply to: [195171609](https://github.com/aspnet/LibraryManager/pull/158#discussion_r195171609) [](ancestors = 195171609)

File: src/libman/Contracts/ConsoleLogger.cs [hunk 8]
```diff
@@ -35,6 +35,29 @@ public string GetUserInput(string fieldName)
...
+        /// 
+        /// </summary>
+        /// <param name="fieldName"></param>
+        /// <param name="defaultValue"></param>
+        /// <returns></returns>
+        public string GetUserInputWithDefault(string fieldName, string defaultValue)
+        {
+            lock (_syncObject)
+            {
+                string message = $"{fieldName} [{defaultValue}]: <enter>";
```
Comment (jimmylewis) [inline src/libman/Contracts/ConsoleLogger.cs]: ><enter> [](start = 65, length = 7)

remove

Comment (ToddGrun) [reply src/libman/Contracts/ConsoleLogger.cs]: :thumbsup: 

---
In reply to: [195176822](https://github.com/aspnet/LibraryManager/pull/158#discussion_r195176822) [](ancestors = 195176822)

**Pull request merged**